multiversx_sdk_core/transaction_factories/smart_contract_transactions_factory.py

```diff
@@ -57,24 +59,83 @@ def create_transaction_for_deploy(self,
 
         return transaction
 
-    def create_transaction_intent_for_execute(self,
-                                              sender: IAddress,
-                                              contract_address: IAddress,
-                                              function: str,
-                                              gas_limit: int,
-                                              arguments: List[Any] = []) -> Transaction:
-        parts = [function] + args_to_strings(arguments)
+    def create_transaction_for_execute(self,
+                                       sender: IAddress,
+                                       contract: IAddress,
+                                       function: str,
+                                       gas_limit: int,
+                                       arguments: List[Any] = [],
+                                       native_transfer_amount: int = 0,
+                                       token_transfers: List[TokenTransfer] = []) -> Transaction:
+        number_of_tokens = len(token_transfers)
+
+        if native_transfer_amount and number_of_tokens:
+            raise BadUsageError("Can't send both native token and ESDT/NFT tokens")
+
+        token_computer = TokenComputer()
+        transfer_args: List[str] = []
+
+        if len(token_transfers) == 1:
+            transfer = token_transfers[0]
+
+            if token_computer.is_fungible(transfer.token):
+                transfer_args = self._build_args_for_esdt_transfer(transfer)
+                transfer_args.append(arg_to_string(function))
+                transfer_args.extend(args_to_strings(arguments))
+            else:
+                transfer_args = self._build_args_for_single_esdt_nft_transfer(transfer)
+                transfer_args.append(contract.hex())
+                contract = sender
+                transfer_args.append(arg_to_string(function))
+                transfer_args.extend(args_to_strings(arguments))
+        elif len(token_transfers) > 1:
+            transfer_args = self._build_args_for_multi_esdt_nft_transfer(contract.hex(), token_transfers)
+            contract = sender
+            transfer_args.append(arg_to_string(function))
+            transfer_args.extend(args_to_strings(arguments))
+        else:
+            transfer_args.append(function)
+            transfer_args.extend(args_to_strings(arguments))
```

Comment:
i think you can refactor this section with something like this, on order to avoid mulitple append, extend of function and arguments:
```
        function_arg = arg_to_string(function)

        if len(token_transfers) == 1:
            transfer = token_transfers[0]

            if token_computer.is_fungible(transfer.token):
                transfer_args = self._build_args_for_esdt_transfer(transfer)
            else:
                transfer_args = self._build_args_for_single_esdt_nft_transfer(transfer)
                transfer_args.append(contract.hex())
                contract = sender
        elif len(token_transfers) > 1:
            transfer_args = self._build_args_for_multi_esdt_nft_transfer(contract.hex(), token_transfers)
            contract = sender
        else:
            function_arg = function

        transfer_args.append(function_arg)
        transfer_args.extend(args_to_strings(arguments))
```

Reply:
Indeed. I realized that the code was not very clean, but I wouldn't change it now, because in the next PR I have refactored it. Here is the PR that will come right after this: https://github.com/multiversx/mx-sdk-py-core/pull/43